MAINT: Simplify scalar power

seberg opened the pull request:
This consolidates the multiple copies of almost identical scalar power
code. It fixes the insignificant fact, that integers were doing some
floating point flag handling, even though our scalar power does not
use that.
The previous code special cased powers of 0, but the C-function seems
to handle it just fine.

---

Well, I was looking into more cleanups, but they are trickier then I thought.  So starting with this.  I want to change of some of the `NotImplemented` paths (and generally some of the calculation as well here), which is mostly unrelated, but I hoped I can make it nicer by cleaning up first.

The current logic is pretty opaque, and often very much not optimal (and strange about what *is* optimized).  I.e. we could optimize only identical types, or we can optimize all types where one can be safely cast to the other (which is fine for the numeric ops).  The current code does the latter, but only half the time: `float64 + float32` is fast, but `float32 + float64` is slow. I suppose that is the more common order, but I doubt its intentional.

seberg committed:
MAINT: Simplify scalar power

This consolidates the multiple copies of almost identical scalar power
code. It fixes the insignificant fact, that integers were doing some
floating point flag handling, even though our scalar power does not
use that.
The previous code special cased powers of 0, but the C-function seems
to handle it just fine.

seberg commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -1058,85 +976,25 @@ static PyObject *
                 "Integers to negative integer powers are not allowed.");
         return NULL;
     }
+#endif
```
With the exception of the signed integers, the math functions all handle 0 input correctly, the only oddity is really the error return.

eric-wieser commented on numpy/core/src/umath/scalarmath.c.src:
```diff
@@ -1058,85 +976,25 @@ static PyObject *
      * here we do the actual calculation with arg1 and arg2
      * as a function call.
      */
+#if @isint@ && !@isuint@
```
These are already mutually exclusive, right? - just `@isint@` would be enough

seberg replied:
ops, I changed it the opposite direction, isint includes uint. The reason is that the float checking is not necessary for all ints here.

Added one more commit, before I forget it, although its unrelated. It just deletes a `volatile` decleration with the comment, since the code now uses `npy_clear_floatstatus_barrier` which solves the same issue as declaring `out` as `volatile`.

seberg committed:
MAINT: Remove volatile and comment from scalar operators

The volatile statement was designed to prevent reordering of
floating point error checks, however, this was more generally
fixed in gh-11036, thus removing the need for the volatile
declaration (and bringing the code in line with the rest of
the file).

mattip commented:
Thanks @seberg
